shtdlib.sh

```diff
@@ -204,17 +204,85 @@ function readlink_m {
     fi
 }
 
+# A function to get the status of shell options supported by 'set'
+# Outputs: 'on', 'off' or nothing for invalid/empty option names
+function get_sh_opts {
+    if [ -n "${1}" ]; then
+        shopt -o "${1}" 2> /dev/null | awk '{print $2}'
+    fi
+}
+
+
+# A function to toggle shell options in bash
+#  - Param 1: [REQUIRED] name of the option (from the 'set' builtin)
+#  - Param 2: [OPTIONAL] 'enable', 'disable' or empty/unset. If empty/unset,
+#             then the option will be toggled
+function toggle_sh_opt {
+    # Make sure enough arguments are passed in
+    args=( ${@} )
+    assert [ ${#args[@]} -gt 0 ]
+    if [ -n "${2}" ]; then
+        if [[ ! ( "${2}" = "enable" || "${2}" = "disable" ) ]]; then
+            color_echo red "Invalid action specified, '${2}'"
+            return
+        fi
+    fi
+
+    # Ensure we're using bash
+    if [ -n "${BASH_VERSION}" ]; then
+        # This is to avoid using a subshell (subprocess preserves shell options)
+        local opt_status
+        mapfile -t opt_status < <(get_sh_opts "${1}")
+
+        if [[ "${opt_status}" = ""  || ! ( "${opt_status}" = "on" || "${opt_status}" = "off" ) ]]; then
+            color_echo red "Invalid option status, '${opt_status}' for option '${1}'"
+            return 1
+        fi
+        debug 10 "Before toggle: '${1}' is set to '${opt_status}'"
+
+        case "${opt_status}" in
+        "on")
+            case "${2}" in
+            "disable")
+                shopt -uo "${1}"
+                debug 10 "After toggle: Unset '${1}'"
+                ;;
+            *)
+                color_echo yellow "No action taken, '${1}' is already set"
+                ;;
+            esac
+            ;;
+        "off")
+            case "${2}" in
+            "enable")
+                shopt -so "${1}"
+                debug 10 "After toggle: Set '${1}'"
+                ;;
+            *)
+                color_echo yellow "No action taken, '${1}' is already unset"
```

Comment:
Since these are library functions we might just want to use debug instead,  and set the level to something like 3, what do you think?

Reply:
K